File: CMPS12B/Dictionary C/Dictionary.c
//-----------------------------------------------------------------------------
//   Dictionary.c
//   Implementation file for IntegerStack ADT
//   Author: Edwin Wang
//   Author: ekwang
//   Assignment: 12L lab5
//   Created on: May 15th, 2019
//-----------------------------------------------------------------------------

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include "Dictionary.h"

// private types --------------------------------------------------------------

// NodeObj
typedef struct NodeObj {
	char* key;
	char* value;
	struct NodeObj* next;
} NodeObj;

// Node
typedef NodeObj* Node;

// newNode()
// constructor of the Node type
Node newNode(char* x, char* y) {
	Node N = malloc(sizeof(NodeObj));
	assert(N!=NULL);
	N->key = x;
	N->value = y;
	N->next = NULL;
	return (N);
}

// freeNode()
// destructor for the Node type
void freeNode(Node* pN) {
	if (pN != NULL && *pN != NULL) {
		free(*pN);
		*pN = NULL;
	}
}

// DictionaryObj
typedef struct DictionaryObj {
	Node head;
	Node tail;
	int numItems;
} DictionaryObj;

// public functions -----------------------------------------------------------

// newStack()
// constructor for the Stack type
Dictionary newDictionary(void) {
	Dictionary D = malloc(sizeof(DictionaryObj));
	assert(D!=NULL);
	D->head = NULL;
	D->tail = NULL;
	D->numItems = 0;
	return D;
}

// freeDictionary()
// destructor for the Dictionary type
void freeDictionary(Dictionary* pD) {
	if (pD != NULL && *pD != NULL) {
		if (!isEmpty(*pD))
			makeEmpty(*pD);
		free(*pD);
		*pD = NULL;
	}
}

// isEmpty()
// returns 1 (true) if D is empty, 0 (false) otherwise
// pre: none
int isEmpty(Dictionary D) {
	if (D == NULL) {
		fprintf(stderr,
				"Dictionary Error: calling isEmpty() on NULL Dictionary reference\n");
		exit(EXIT_FAILURE);
	}
	return (D->numItems == 0);
}

// size()
// returns the number of (key, value) pairs in D
// pre: none
int size(Dictionary D) {
	return (D->numItems);
}

// lookup()
// returns the value v such that (k, v) is in D, or returns NULL if no
// such value v exists.
// pre: none
char* lookup(Dictionary D, char* k) {
	for (Node N = D->head; N != NULL; N = N->next)
		if (!strcmp(k, N->key))
			return N->value;
	return NULL;
}

// insert()
// inserts new (key,value) pair into D
// pre: lookup(D, k)==NULL
void insert(Dictionary D, char* k, char* v) {
	if (lookup(D, k) != NULL) {
		fprintf(stderr,
				"Dictionary Error: insert() called with existing key\n");
		exit(EXIT_FAILURE);
	}
	if (D->numItems == 0) { // No items in Dictionary
		D->tail = newNode(k, v);
		D->head = D->tail;
	} else { // Items already in Dictionary
		D->tail->next = newNode(k, v);
		D->tail = D->tail->next;
	}
	++D->numItems;
}

// delete()
// deletes pair with the key k
// pre: lookup(D, k)!=NULL
void delete(Dictionary D, char* k) {
	if (lookup(D, k) == NULL) {
		fprintf(stderr,
				"Dictionary Error: calling delete() on non-existing pair\n");
		exit(EXIT_FAILURE);
	}
	--D->numItems;
	Node N = D->head;
	Node currentNode = D->head;
	if (D->numItems == 0) { // Delete if target is only Node in Dictionary
		D->head = NULL;
		D->tail = NULL;
	} else if (!strcmp(D->head->key, k)) { // Delete if target Node is head
		D->head = N->next;
	} else if (!strcmp(D->tail->key, k)) { // Delete if target Node is tail
		for (; strcmp(currentNode->next->key, k);
				currentNode = currentNode->next)
			;
		N = currentNode->next;
		currentNode->next = NULL;
		D->tail = currentNode;
	} else { // Delete if target in middle
		for (; strcmp(currentNode->next->key, k);
				currentNode = currentNode->next)
			;
		N = currentNode->next;
		currentNode->next = currentNode->next->next;
	}
	freeNode(&N);
}

// pop()
// deletes item at top of D
// pre: !isEmpty(S)
void pop(Dictionary D) {
	Node N;
	if (D == NULL) {
		fprintf(stderr,
				"Dictionary Error: calling pop() on NULL Dictionary reference\n");
		exit(EXIT_FAILURE);
	}
	if (D->numItems == 0) {
		fprintf(stderr,
				"Dictionary Error: calling pop() on empty Dictionary\n");
		exit(EXIT_FAILURE);
	}
	N = D->head;
	D->head = D->head->next;
	--D->numItems;
	freeNode(&N);
}

// makeEmpty()
// resets S to the empty state
// pre: none
void makeEmpty(Dictionary D) {
	while (D->numItems > 0)
		pop(D);
	D->tail = NULL;
	D->head = NULL;
}

// printDictionary()
// prints a text representation of S to the file pointed to by out
// pre: none
void printDictionary(FILE* out, Dictionary D) {
	Node N = D->head;
	if (D == NULL) {
		fprintf(stderr,
				"Dictionary Error: calling printDictionary() on NULL Dictionary reference\n");
		exit(EXIT_FAILURE);
	}
	for (int i = 0; i < D->numItems; ++i) {
		fprintf(out, "%s %s\n", N->key, N->value);
		N = N->next;
	}
}

File: CMPS12B/Dictionary/Dictionary.java
/*
 * Name: Edwin Wang
 * CruzID: ekwang
 * Assignment: 12B pa3
 * Date: 4/30/2019
 * Description: Dictionary ADT
 */

public class Dictionary implements DictionaryInterface {

	// private inner Node class
	private class Node {
		String key;
		String value;
		Node next;

		Node(String k, String v) {
			key = k;
			value = v;
			next = null;
		}
	}

	// Fields for the Dictionary class
	private Node head; // reference to first Node in List
	private Node last; // reference to last Node in List
	private int numItems; // number of items in this Dictionary

	// Dictionary()
	// constructor for the Dictionary class
	public Dictionary() {
		head = null;
		last = null;
		numItems = 0;
	}

	// isEmpty()
	// pre: none
	// returns true if this Dictionary is empty, false otherwise
	@Override
	public boolean isEmpty() {
		if (numItems == 0) {
			return true;
		}
		return false;
	}

	// size()
	// pre: none
	// returns the number of entries in this Dictionary
	@Override
	public int size() {
		return numItems;
	}

	// lookup()
	// pre: none
	// returns value associated key, or null reference if no such key exists
	@Override
	public String lookup(String key) {
		Node node = findKey(key);
		if (node == null)
			return null;
		return node.value;
	}

	// insert()
	// inserts new (key,value) pair into this Dictionary
	// pre: lookup(key)==null
	@Override
	public void insert(String key, String value) {
		Node node = new Node(key, value);
		if (lookup(key) != null) {
			throw new DuplicateKeyException();
		}
		if (numItems == 0) {
			head = node;
			last = node;
		} else {
			last.next = node;
			last = node;
		}
		++numItems;
	}

	// delete()
	// deletes pair with the given key
	// pre: lookup(key)!=null
	@Override
	public void delete(String key) {
		Node node = head;
		if (lookup(key) == null)
			throw new KeyNotFoundException();
		else {
			if (head.key.equals(key))
				head = head.next;
			else
				while (node.next != null) {
					if (node.next.key.equals(key)) {
						if (last == node.next) {
							last = node;
						}
						node.next = node.next.next;
						break;
					}
					node = node.next;
				}
			--numItems;
		}
	}

	// makeEmpty()
	// pre: none
	@Override
	public void makeEmpty() {
		head = null;
		last = null;
		numItems = 0;
	}

	private Node findKey(String key) { // return reference to Node with key, or null otherwise
		Node current = head;
		while (current != null) {
			if (current.key.equals(key)) {
				return current;
			}
			current = current.next;
		}
		return null;
	}

	// toString()
	// returns a String representation of this Dictionary
	// overrides Object's toString() method
	// pre: none
	public String toString() {
		String output = new String("");
		Node node = head;
		while (node != null) {
			output += node.key + " " + node.value + "\n";
			node = node.next;
		}
		return output;
	}

}
